From 95bb06b3a42c4a371e6a491da594c9d33754d1a3 Mon Sep 17 00:00:00 2001
From: He jiean <hejiean@eswincomputing.com>
Date: Wed, 9 Jul 2025 16:43:28 +0800
Subject: [PATCH 618/654] fix: add dmabuf_helper params check

Changelogs:
1. make sure all heap_obj returen value as ERR_PTR when error occurs.
2. all export functions check if heap obj is error pointer when heap_obj
is input parameter.
3. add comment to header, explain the return value be NULL or error
pointer.
4. make sure all heap obj return value check using IS_ERR in user
drivers.

Signed-off-by: hejiean <hejiean@eswincomputing.com>
---
 .../dmabuf-heap-import-helper.c               | 25 +++++++++++++------
 .../staging/media/eswin/dewarp/dw200_ioctl.c  |  4 +--
 include/linux/dmabuf-heap-import-helper.h     | 14 +++++++++++
 3 files changed, 33 insertions(+), 10 deletions(-)

diff --git a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
index c4ce2e01a61f..c634da8df58e 100644
--- a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
+++ b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
@@ -342,7 +342,7 @@ struct heap_mem *common_dmabuf_lookup_heapobj_by_fd(struct heap_root *root, int
 	/* get dmabuf handle */
 	dma_buf = dma_buf_get(fd);
 	if (IS_ERR(dma_buf))
-		return NULL;
+		return ERR_CAST(dma_buf);
 
 	mutex_lock(&root->lock);
 	ret = dmabuf_heap_lookup_buf_handle(&root->fp, dma_buf, (uint64_t *)&heap_obj);
@@ -352,7 +352,7 @@ struct heap_mem *common_dmabuf_lookup_heapobj_by_fd(struct heap_root *root, int
 	if (0 == ret)
 		return heap_obj;
 	else
-		return NULL;
+		return ERR_PTR(ret);
 }
 EXPORT_SYMBOL(common_dmabuf_lookup_heapobj_by_fd);
 
@@ -370,7 +370,7 @@ struct heap_mem *common_dmabuf_lookup_heapobj_by_dma_buf_st(struct heap_root *ro
 	if (0 == ret)
 		return heap_obj;
 	else
-		return NULL;
+		return ERR_PTR(ret);
 }
 EXPORT_SYMBOL(common_dmabuf_lookup_heapobj_by_dma_buf_st);
 
@@ -413,8 +413,12 @@ static void __common_dmabuf_heap_release(struct kref *kref)
 
 void common_dmabuf_heap_release(struct heap_mem *heap_obj)
 {
-	struct heap_root *root = heap_obj->root;
+	struct heap_root *root;
 
+	if (WARN_ON(IS_ERR(heap_obj)))
+		return;
+
+	root = heap_obj->root;
 	mutex_lock(&root->lock);
 	kref_put(&heap_obj->refcount, __common_dmabuf_heap_release);
 	mutex_unlock(&root->lock);
@@ -426,8 +430,7 @@ void *common_dmabuf_heap_map_vaddr(struct heap_mem *heap_obj)
     struct dma_buf_map map;
     int ret;
 
-	WARN_ON(!heap_obj);
-	if (!heap_obj)
+	if (WARN_ON(IS_ERR(heap_obj)))
 		return NULL;
 
     if (heap_obj->vaddr)
@@ -448,7 +451,9 @@ void common_dmabuf_heap_umap_vaddr(struct heap_mem *heap_obj)
 {
 	struct dma_buf_map map;
 
-	WARN_ON(!heap_obj);
+	if (WARN_ON(IS_ERR(heap_obj)))
+		return;
+
 	if (heap_obj && heap_obj->vaddr) {
         map.vaddr = heap_obj->vaddr;
         map.is_iomem = 0;
@@ -584,8 +589,12 @@ static void __common_dmabuf_heap_rsv_iova_unmap(struct kref *kref)
 
 void common_dmabuf_heap_rsv_iova_unmap(struct heap_mem *heap_obj)
 {
-	struct heap_root *root = heap_obj->root;
+	struct heap_root *root;
+
+	if (WARN_ON(IS_ERR(heap_obj)))
+		return;
 
+	root = heap_obj->root;
 	mutex_lock(&root->lock);
 	kref_put(&heap_obj->refcount, __common_dmabuf_heap_rsv_iova_unmap);
 	mutex_unlock(&root->lock);
diff --git a/drivers/staging/media/eswin/dewarp/dw200_ioctl.c b/drivers/staging/media/eswin/dewarp/dw200_ioctl.c
index 0088cf5704e8..b28665ba1f3b 100644
--- a/drivers/staging/media/eswin/dewarp/dw200_ioctl.c
+++ b/drivers/staging/media/eswin/dewarp/dw200_ioctl.c
@@ -179,7 +179,7 @@ static u64 get_dma_addr(struct dw200_subdev *dev, int dmabuf_fd,
 	u64 dma_addr;
 
 	hmem = common_dmabuf_lookup_heapobj_by_fd(dev->pheap_root, dmabuf_fd);
-	if (!hmem) {
+	if (IS_ERR(hmem)) {
 		pr_err("Could not find dmabuf-heap for dmabuf fd %d, Please import it first\n",
 		       dmabuf_fd);
 		return -1;
@@ -226,7 +226,7 @@ int dma_heap_iova_release(struct dw200_subdev *dev, int dmabuf_fd)
 	}
 
 	hmem = common_dmabuf_lookup_heapobj_by_fd(dev->pheap_root, dmabuf_fd);
-	if (!hmem) {
+	if (IS_ERR(hmem)) {
 		pr_err("Cannot find dmabuf-heap for dmabuf_fd %d\n", dmabuf_fd);
 		return -1;
 	}
diff --git a/include/linux/dmabuf-heap-import-helper.h b/include/linux/dmabuf-heap-import-helper.h
index d4acd7c5177c..913134b88cc8 100644
--- a/include/linux/dmabuf-heap-import-helper.h
+++ b/include/linux/dmabuf-heap-import-helper.h
@@ -114,20 +114,34 @@ static inline void common_dmabuf_heap_set_dir(struct heap_mem *heap_obj, enum dm
 void common_dmabuf_heap_import_init(struct heap_root *root, struct device *dev);
 void common_dmabuf_heap_import_uninit(struct heap_root *root);
 
+/* lookup heapobj from heap tree by fd, return heapobj or ERR pointer. */
 struct heap_mem *common_dmabuf_lookup_heapobj_by_fd(struct heap_root *root, int fd);
+
+/* lookup heapobj from heap tree by dma_buf, return heapobj or ERR pointer. */
 struct heap_mem *common_dmabuf_lookup_heapobj_by_dma_buf_st(struct heap_root *root, struct dma_buf *dma_buf);
 
+/* import a fd binded dma buf to heap root, return heap obj or ERR_PTR  */
 struct heap_mem *common_dmabuf_heap_import_from_user(struct heap_root *root, int fd);
+
+/* import dma buf to heap root, return heap obj or ERR_PTR */
 struct heap_mem *common_dmabuf_heap_import_from_user_with_dma_buf_st(struct heap_root *root, struct dma_buf *dma_buf);
+
+/* release the heap obj, auto unmap the binded vaddr */
 void common_dmabuf_heap_release(struct heap_mem *heap_obj);
 
+/* map a virtual address to visit the heap obj, return address Or NULL */
 void *common_dmabuf_heap_map_vaddr(struct heap_mem *heap_obj);
+
+/* unmap the virtual address binded in heap obj */
 void common_dmabuf_heap_umap_vaddr(struct heap_mem *heap_obj);
 
+/* alloc a dma buf and bind to heap root, return heap obj or ERR_PTR */
 struct heap_mem *common_dmabuf_heap_import_from_kernel(struct heap_root *root, char *name, size_t len, unsigned int fd_flags);
 
+/* split a dma buf, return new fd or error */
 int esw_common_dmabuf_split_export(int dbuf_fd, unsigned int offset, size_t len, int fd_flags, char *name);
 
+/* map a virtual address to visit the heap obj, return address Or NULL */
 struct heap_mem *common_dmabuf_heap_rsv_iova_map(struct heap_root *root, int fd, dma_addr_t iova, size_t size);
 void common_dmabuf_heap_rsv_iova_unmap(struct heap_mem *heap_obj);
 void common_dmabuf_heap_rsv_iova_uninit(struct heap_root *root);
-- 
2.51.0

